packages/pg-gateway/src/connection.ts

```diff
@@ -275,353 +279,301 @@ export default class PostgresConnection {
-            const password = this.reader.cstring();
-
-            // We must pause/resume the socket before/after each hook to prevent race conditions
-            this.socket.pause();
-            const valid = await this.options.validateCredentials?.(
-              {
-                authMode,
-                user: this.clientInfo.parameters.user,
-                password,
-              },
-              this.state
-            );
-            this.socket.resume();
-
-            if (!valid) {
-              this.sendAuthenticationFailedError();
-              this.socket.end();
-              return;
-            }
-
-            await this.completeAuthentication();
-            return;
-          }
-          case 'md5Password': {
-            const hash = this.reader.cstring();
-            const valid = await this.options.validateCredentials?.(
-              {
-                authMode,
-                user: this.clientInfo.parameters.user,
-                hash,
-                salt: this.md5Salt,
-              },
-              this.state
-            );
-
-            if (!valid) {
-              this.sendAuthenticationFailedError();
-              this.socket.end();
-              return;
-            }
-
-            await this.completeAuthentication();
-            return;
-          }
-        }
-        return;
-      }
-      case FrontendMessageCode.Query: {
-        const { query } = this.readQuery();
+    if (!isValid) {
+      this.sendError({
+        severity: 'FATAL',
+        code: '08000',
+        message: 'client certificate is invalid',
+      });
+      this.socket.end();
+      return;
+    }
 
-        console.log(`Query: ${query}`);
+    return true;
+  }
 
-        // TODO: call `onQuery` hook to allow consumer to choose how queries are implemented
+  private async handleRegularMessage(message: Buffer): Promise<void> {
+    const code = this.reader.byte();
 
+    switch (code) {
+      case FrontendMessageCode.Terminate:
+        this.handleTerminate(message);
+        break;
+      default:
         this.sendError({
           severity: 'ERROR',
           code: '123',
-          message: 'Queries not yet implemented',
+          message: 'Message code not yet implemented',
         });
         this.sendReadyForQuery('idle');
-        return;
-      }
-      // @see https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-TERMINATION
-      case FrontendMessageCode.Terminate: {
-        this.socket.end();
-        return;
-      }
     }
   }
 
+  handleTerminate(message: Buffer) {
+    this.socket.end();
+  }
+
+  private isSslRequest(message: Buffer): boolean {
+    return message.length === 8 && message.readInt32BE(4) === 80877103;
+  }
+
+  private isStartupMessage(message: Buffer): boolean {
+    // StartupMessage begins with length (Int32) followed by protocol version (Int32)
+    return message.length > 8 && message.readInt32BE(4) === 196608; // 196608 is protocol version 3.0
+  }
```

Comment:
Even though the PG docs mention the decimal `196608` I think it's clearer if we parse it as two 16 bytes (like we do elsewhere) to pull out the 3 and the 0:

```suggestion
  private isStartupMessage(message: Buffer): boolean {
    // StartupMessage begins with length (Int32) followed by major version (Int16) and minor version (Int16)
    const majorVersion = message.readInt16BE(4);
    const minorVersion = message.readInt16BE(6);
    return message.length > 8 && majorVersion === 3 && minorVersion === 0; // protocol version 3.0
  }
```

Reply:
Thanks, it's clearer. I also changed the logic of isSslRequest to be closer to the docs. 
Fixed in https://github.com/supabase-community/pg-gateway/pull/1/commits/1a641dfaea1ad3483cf3521170a2e653ccaeefbb